feat(#1381): Add a way to specify "inject-only" with @JacksonInject

giulong opened the pull request:
(implements #1381)

useInput=TRUE in JacksonInject now make deserialization discard the injected value in favor of input (if any)

giulong committed:
feat(#1381): useInput=TRUE in JacksonInject now make deserialization discard the injected value in favor of input (if any)

giulong commented on src/test/java/com/fasterxml/jackson/databind/deser/inject/JacksonInject2678Test.java:
```diff
@@ -1,4 +1,4 @@
-package com.fasterxml.jackson.databind.tofix;
+package com.fasterxml.jackson.databind.deser.inject;
```
moved since this should not fail anymore

giulong commented on src/test/java/com/fasterxml/jackson/databind/deser/inject/JacksonInject1381Test.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    @Test
+    @DisplayName("input YES, injectable YES, useInput DEFAULT|FALSE => injected")
```
According to `useInput`'s [javadocs](https://github.com/FasterXML/jackson-annotations/blob/2.x/src/main/java/com/fasterxml/jackson/annotation/JacksonInject.java#L34):
> Default is `OptBoolean.DEFAULT`, which translates to `OptBoolean.TRUE`: this is
   for backwards compatibility (2.8 and earlier always allow binding input value).

This combination: `input YES, injectable YES, useInput DEFAULT` should actually behave as if we had `useInput = TRUE`, that is dropping the injected value and returning the input.

Nevertheless, since `useInput` never worked that way, this would mean introducing breaking changes. My suggestion is to keep things as per this test and change the javadoc accordingly:
> Default is `OptBoolean.DEFAULT`, which translates to `OptBoolean.FALSE`

giulong commented on src/test/java/com/fasterxml/jackson/databind/deser/inject/JacksonInject1381WithOptionalTest.java:
```diff
@@ -0,0 +1,115 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class JacksonInject1381WithOptionalTest extends DatabindTestUtil {
```
This class shows how the behavior changes for all the combinations of the other test class when all injected fields are optional. Do you think we still need a `required` property in the annotation?

I feel like combining `useInput` and `optional` would cover all cases, also because `optional` should drive the same logic (but reversed) of `required`. Given this, adding a `required` property would make things unclear for the users, and the internal logic would be quite messy. Does this make sense?

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerFactory.java:
```diff
@@ -817,11 +817,22 @@ protected void addInjectables(DeserializationContext ctxt,
+                    useInput = injectableValue.getUseInput();
+                }
+
+                if (!Boolean.TRUE.equals(useInput)) {
```
I think this is slightly off: Javadoc for `useInput` says:

    Default is `OptBoolean.DEFAULT`, which translates to `OptBoolean.TRUE`.

I can change this.

cowtowncoder replied:
Oh. Never mind, later comment spells it out

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder commented on src/test/java/com/fasterxml/jackson/databind/deser/inject/JacksonInject1381Test.java:
```diff
@@ -0,0 +1,123 @@
+
+    @Test
+    @DisplayName("input NO, injectable YES, useInput DEFAULT|TRUE|FALSE => injected")
+    void test2() throws JsonProcessingException {
```
Let's not use "throws JsonProcessingException" since merging to 3.0 gets tricky (base is `JacksonException`, JPE remove). Instead just use plain "throws Exception".

cowtowncoder replied:
done.

cowtowncoder committed:
Change exception throws

cowtowncoder committed:
...

cowtowncoder reviewed (approved):
LGTM.

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerFactory.java:
```diff
@@ -817,11 +817,23 @@ protected void addInjectables(DeserializationContext ctxt,
+
+                // 04-Jun-2025, tatu: [databind#1381]: default for "useInput" is false
+                if (!Boolean.TRUE.equals(useInput)) {
+                    builder.addInjectable(PropertyName.construct(m.getName()),
```
This is part I don't fully understand: if injectable is only added if no input is to be used, how do things work with `useInput = false`, injectable value being present.

giulong replied:
As per this test, with no input but injected value present, the injected value is always returned regardless of `useInput`:
```java
@Test
@DisplayName("input NO, injectable YES, useInput DEFAULT|TRUE|FALSE => injected")
void test2() throws Exception {
    assertEquals("injected", injectedMapper.readValue(empty, InputDefault.class).getField());
    assertEquals("injected", injectedMapper.readValue(empty, InputTrue.class).getField());
    assertEquals("injected", injectedMapper.readValue(empty, InputFalse.class).getField());
}
```
I read it like "we have no input but we have a value to inject, so we don't care about anything else: let's use the injected value". Is this the expected behavior? Or you meant something else?

cowtowncoder replied:
Logic-wise, yes, you are correct (expected behavior I agree with).

But I meant code; meaning...  what  does `addInjectable()` do -- to me, it's the thing that injects non-input (injectable) value. So why is that contingent on "useInput"? And I think check here seems wrong in that sense: injectable value may be needed even input is used.
It is not only injected if `useInput` is false (or missing); can also be injected if `useInput` is `true` but input has no value for property.

giulong replied:
It's a way to eagerly discard the injectable if we're going to use the input anyways. I think it's better in terms of performance since any further evaluation is useless, and there are many methods involved down that path.
But I also get this:
> injectable value may be needed even input is used

Not now maybe, but in general I get that if a field is marked as injected, we should add it to the internal injectables for whatever future logic.

I can try to change this, which means passing `useInput` to the builder (last param in the snippet), as we made for `optional`:
```java
builder.addInjectable(PropertyName.construct(m.getName()),
                            m.getType(),
                            beanDesc.getClassAnnotations(), m, entry.getKey(), optional, useInput);
```
Maybe we could also think of passing the whole `injectableValue` to the builder, instead of passing its properties one by one.

But I think this is not going to be easy. What do you think, is it worth it?

cowtowncoder replied:
Yes, figured it was meant as optimization.
But the problem with " if we're going to use the input anyways" is that there might not be any input.
So it's not "use input if any matching, otherwise `null`" but rather "use input if present; if not, use injected value". Otherwise why even mark it with `@JacksonInject` at all, actually.

So we cannot know statically which one to use.

And yes, I think this is necessary to handle properly.

But I understand things are messy, tricky, complicated, esp. when passing things via Constructors (CreatorProperty properties) vs Field/MethodProperty ones.

cowtowncoder reviewed (dismissed):
Actually, no: if I change things to put `@JacksonInject` into Constructor parameter (instead of Field) -- common alternative (although adding on Field is allowed too), a few tests fail.

So this doesn't quite work fully yet.

cowtowncoder committed:
Add release notes; broke tests by changing annotation placements

cowtowncoder commented:
Ok. I don't think this fix is complete: I changed test set up slightly (in a way I think users are more likely to use it -- although original way is legit too... but both should work) and now a few tests fail.

So would need to figure a way to make things fully work with combination of input and/or injectable values.

giulong commented:
Fair point, I missed that constructor properties take another path. On it

giulong committed:
test: minor refactor to apply coding conventions

giulong committed:
test(#1381): covering all cases for JacksonInject, whether we useInput, optional, and/or the DeserializationFeature

giulong committed:
feat(#1381): JacksonInject.useInput works for constructor properties as well

giulong committed:
Merge branch '2.x' into feature/databind-1381-inject-only

giulong commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/PropertyValueBuffer.java:
```diff
@@ -218,6 +219,25 @@ public Object[] getParameters(SettableBeanProperty[] props)
             _creatorParameters[_anyParamSetter.getParameterIndex()] = _createAndSetAnySetterValue();
         }
+
+        for (int ix = 0; ix < props.length; ++ix) {
```
This is the moment when creator properties are resolved, so I think it's ok to have the logic here, as per the method's javadoc as well:
> /**
     * Method called to do necessary post-processing such as injection of values
     * and verification of values for required properties,
     * after either {@link #assignParameter(SettableBeanProperty, Object)}
     * returns <code>true</code> (to indicate all creator properties are found), or when
     * then whole JSON Object has been processed,
     */
    public Object[] getParameters(SettableBeanProperty[] props)

I placed the new part right before the code that checks for null creator properties in case we have an injected value to use, so to not fail. Should I move it after?

cowtowncoder replied:
It makes sense to me to do it after, so injection can prevent failure.

cowtowncoder replied:
However: this does another pass, as earlier calls to `_findMissing()` -- only in case of no input, tho -- has already done lookups. So we'd rather avoid those calls.

I'll add a high-level note on this, I think we can avoid redundancy.

giulong commented:
We now have the tests to cover all the scenarios (I guess), meaning combinations of injected creator properties vs fields, with or without `JacksonInject.optional()` and `DeserializationFeature.FAIL_ON_UNKNOWN_INJECT_VALUE`. So, in case other changes are needed, at least we have a no regression suite to check them.

cowtowncoder commented:
> We now have the tests to cover all the scenarios (I guess), meaning combinations of injected creator properties vs fields, with or without `JacksonInject.optional()` and `DeserializationFeature.FAIL_ON_UNKNOWN_INJECT_VALUE`. So, in case other changes are needed, at least we have a no regression suite to check them.

I will try to have a good look tomorrow, to answer with proper understanding.

Thank you for going through this again.

giulong committed:
Merge branch 'FasterXML:2.x' into feature/databind-1381-inject-only

giulong committed:
test(#1381): minor code cleaning removing unused getters

giulong committed:
feat(#1381): always adding properties annotated with JacksonInject to the injectables

giulong commented:
I reworked the logic to always add the properties annotated with `@JacksonInject` to the injectables, regardless of their `optional` and `useInput` values.

Few things to note, for completeness:

1. I'm doing comparisons like `value == JacksonInject.Value.empty()`. I think it's better to not use `equals` here, since we're looking for that specific constant.
2. In [this case](https://github.com/giulong/jackson-databind/blob/feature/databind-1381-inject-only/src/test/java/com/fasterxml/jackson/databind/deser/inject/JacksonInject1381DeserializationFeatureDisabledTest.java#L117-L134):

    ```java
    @Test
    @DisplayName("FAIL_ON_UNKNOWN_INJECT_VALUE NO, input NO, injectable NO, useInput DEFAULT|TRUE|FALSE => exception")
    void test1() {
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputDefault.class));
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputDefaultConstructor.class));
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputTrue.class));
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputTrueConstructor.class));
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputFalse.class));
        assertThrows(ValueInstantiationException.class,
                () -> plainMapper.readValue(empty, InputFalseConstructor.class));
    }
    ```
    
    we expect an exception, since we have no input nor injectable value. Since `FAIL_ON_UNKNOWN_INJECT_VALUE` is disabled as well, the injection process does not throw the usual `MissingInjectableValueExcepion`. The `ValueInstantiationException` comes from the [AnnotatedConstructor.call](https://github.com/giulong/jackson-databind/blob/feature/databind-1381-inject-only/src/main/java/com/fasterxml/jackson/databind/introspect/AnnotatedConstructor.java#L124-L127) method, since the arg provided is actually `JacksonInject.Value.empty()`.
    In the injection process, we use that value as a way to signal that we tried to resolve the injectable, but due to `useInput`, `optional` and `FAIL_ON_UNKNOWN_INJECT_VALUE` the injection should actually be skipped. I think this is fine since we need a special marker (`null` isn't good since it could be a valid value), but from the user's perspective I also think throwing a generic `ValueInstantiationException` is not the best: seeing a message like `Cannot construct instance of ..., problem: argument type mismatch` could be misleading.
3. This implementation worsens #4218, making the injection process happen thrice. Checked with `JacksonInject4218Test`.

I hope the overall implementation is good enough to be merged, despite the bullets above. What do you think?

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Add back accidentally removed CREDITS line (merge)

cowtowncoder committed:
...

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/DeserializationContext.java:
```diff
@@ -465,21 +465,24 @@ public final boolean hasSomeOfFeatures(int featureMask) {
      */
     public final Object findInjectableValue(Object valueId,
-            BeanProperty forProperty, Object beanInstance, Boolean optional)
+            BeanProperty forProperty, Object beanInstance, Boolean optional, Boolean useInput)
```
Just noticed: this is backwards-compatibility breaking change; need to retain old version that just delegates to it (in case something from outside databind calls it).

**EDIT** N/m, continues on earlier change that does delegate as expected.

giulong replied:
The `optional` parameter was added in #5131. We had the same discussion that ended up deprecating and delegating as you can see [here](https://github.com/FasterXML/jackson-databind/pull/5131/files#diff-25c6303c76a7f05a72cdc35a1271bcc9374db3745d61ab895e7199f050619f7e).

I didn't do the same thing again here since we're talking about the same version (2.20) which is not yet released, so I guess it's safe since no one is yet using the `optional` parameter, except for snapshots maybe. Is it ok?

cowtowncoder replied:
@giulong You are correct: I only realized this after adding my comment. So nothing to change here, it's all good.

I'll be out for the weekend but hope to pick this up on monday at latest, to see if this can go in 2.20.0 final.

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/InjectableValues.java:
```diff
@@ -28,7 +28,7 @@ public abstract class InjectableValues
     public Object findInjectableValue(DeserializationContext ctxt,
             Object valueId, BeanProperty forProperty, Object beanInstance,
-            Boolean optional)
+            Boolean optional, Boolean useInput)
```
Same here, incompatible change.

**EDIT**: or maybe not, due to earlier change. But should add `@since 2.20` tag (which should have been added earlier but seems to have been missed). Will actually add.

cowtowncoder replied:
Will resolve, please ignore.

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder commented:
Ok: I merged plumbing changes (to pipe through `useInput`). There's now one test fail wrte concurrently fixed (but now broken) #4218.

Will go over remaining PR again, will see if there's a chance this could make it in 2.20.0 or not.

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Test cleanup for easier 3.0 merging

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/SettableBeanProperty.java:
```diff
@@ -468,6 +469,14 @@ public int getCreatorIndex() {
+     *
+     * @since 2.20
+     */
+    public JacksonInject.Value getInjectableValue() { return null; }
```
Will rename: it's not actual injectable value but definition of such thing.

cowtowncoder committed:
Minor renaming

cowtowncoder commented:
Ok, made some minor changes but I really need to read through the changes, with focus and thought. This is quite tricky.

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/DeserializationContext.java:
```diff
@@ -471,8 +471,10 @@ public final Object findInjectableValue(Object valueId,
-                    || (optional == null && !isEnabled(DeserializationFeature.FAIL_ON_UNKNOWN_INJECT_VALUE))) {
+            if (Boolean.TRUE.equals(useInput)
+                    || Boolean.TRUE.equals(optional)
+                    || (useInput == null || optional == null)
```
This looks odd; should second `||` actually be `&&`? Or maybe it's just odd grouping of clauses with parenthesis.

Should probably update comment to describe logic fully.

giulong replied:
We're returning `empty()` (we want to discard the injected value) in one of these cases:

- either `useInput` or `optional` is `true`
- either `useInput` or `optional` is `null` **and** `FAIL_ON_UNKNOWN_INJECT_VALUE` is disabled

> maybe it's just odd grouping of clauses with parenthesis

Yup, due to formatting, I think the precedence in the last two lines is not immediately clear
```java
if (Boolean.TRUE.equals(useInput)
        || Boolean.TRUE.equals(optional)
        || (useInput == null || optional == null)
        && !isEnabled(DeserializationFeature.FAIL_ON_UNKNOWN_INJECT_VALUE)) {
```

I'm updating the comment

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/ValueInjector.java:
```diff
@@ -69,7 +69,14 @@ public void inject(DeserializationContext context, Object beanInstance)
+        if (value == JacksonInject.Value.empty()) {
+            if (Boolean.FALSE.equals(_optional)) {
+                throw context.missingInjectableValueException(String.format(
+                        "No 'injectableValues' configured, cannot inject value with id '%s'", _valueId),
```
This is misleading: there may be (other) `injectableValues`, but not specific injectable value being looked for.

giulong replied:
Will replace with the same message found in [InjectableValues.java](https://github.com/FasterXML/jackson-databind/blob/2.x/src/main/java/com/fasterxml/jackson/databind/InjectableValues.java#L106-L109):

> "No injectable value with id '%s' found (for property '%s')"

cowtowncoder replied:
Otherwise great, but there's no property name available so this:

```
String.format("No injectable value with id '%s' found (for property '%s')",
    _valueId, beanInstance),
```

is bit wrong (beanInstance is actual POJO). I'll fix that.

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/ValueInjector.java:
```diff
@@ -69,7 +69,14 @@ public void inject(DeserializationContext context, Object beanInstance)
-        if (!JacksonInject.Value.empty().equals(value)) {
+
+        if (value == JacksonInject.Value.empty()) {
+            if (Boolean.FALSE.equals(_optional)) {
```
I think this is slightly wrong as it won't match `null`; and as per `@JacksonInject.optional` Javadocs, default value, expressed as `null` here, is same as `false`.
So I think this should be:

    if (!Boolean.TRUE.equals(_optional)) {

giulong replied:
Yes, but I think it's more complicated than just flipping that condition, since now `optional` must work together with `useInput`. As an example, if I just apply your suggestion, [this test](https://github.com/FasterXML/jackson-databind/pull/5175/files/1e8578d127afb7cb9a2db1564e9a414c81a79982#diff-afa7ec9ead25622b3526ec742a7498af3d9e296f74a577a684aa376838d48321R170) fails:
```java
@DisplayName("input YES, injectable NO, useInput TRUE => input")
void test4() throws Exception {
    assertEquals("input", plainMapper.readValue(input, InputTrue.class).getField());
```
In this case, I think that even if `optional` is `null`, `useInput` should take precedence: the user should not be forced to specify also `optional = true`imho, since they're already saying "use the input whatever the injection is". Forcing users to also specify `optional = true` seems a bit redundant to me. What do you think?

On the other side, there are a couple test that are currently passing but should not, such as [this](https://github.com/FasterXML/jackson-databind/pull/5175/files/1e8578d127afb7cb9a2db1564e9a414c81a79982#diff-8c4463bb6b0fc31654d52d236645e79e18e4bf77c4449084dc176b0694a9408eR151) (the `@DisplayName` is correct while the assertion is wrong):
```java
@DisplayName("FAIL_ON_UNKNOWN_INJECT_VALUE NO, input YES, injectable NO, useInput DEFAULT|FALSE => exception")
    void test3() throws Exception {
        assertEquals("input", plainMapper.readValue(input, InputDefault.class).getField());
```
Here we have:
* an **input**
* no **injection**
* both `optional` and `useInput` are `null`

So I think the user here should explicitly state that `optional = true` or expect an exception. Am I right?

If we agree on the behavior I can start fixing the current implementation.

cowtowncoder replied:
Ok, I am not sure. The problem wrt "input" is simply this: existence of value from input is unknown at processing time (the way things are; not a theoretical limit but practical implementation limit).
So whereas we can easily see if injectable value exists, we can't say -- at time we need it -- whether there is input value. At least when injecting a Field (for Constructor injection we do actually keep track).

Having said that: on

> In this case, I think that even if optional is null, useInput should take precedence: the user should not be forced to specify also optional = true
> imho, since they're already saying "use the input whatever the injection is". Forcing users to also specify optional = true seems a bit redundant to me.
> What do you think?

I think I agree: if "useInput" is effectively `true` (`OptBoolean.TRUE` or default, i.e. not specified), then only explicit `optional = OptionalBoolean.FALSE`) should lead to failure.
My only concern there is slight deviation from existing handling but... that may be more theoretical than actual.

So, yes: I think you are right, after all -- esp. when considering both "optional" and "useInput" together.

cowtowncoder commented:
Ok, so this getting closer!

But I think duplicate setting and traversal could and should be avoided.

So the existing problem is that before this PR injection only occurs in cases where input does not have property value. PR fixes this by iterating over all creator properties and checks each one, injecting as appropriate, possibly re-setting already injected value.

I think we should be able to do better and handle injection case when input is being bound, instead; and blocking input from being used if configured with `useInput = false`.

giulong committed:
Merge branch '2.x' into feature/databind-1381-inject-only

giulong committed:
refactor(#1381): changing exception message in ValueInjector when no value with the provided id is found

giulong committed:
chore(#1381): minor comment update

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

giulong committed:
fix(#1381): avoid duplicate injection on constructor properties

giulong commented:
> There's now one test fail wrte concurrently fixed (but now broken) https://github.com/FasterXML/jackson-databind/issues/4218.

I fixed it with [this commit](https://github.com/FasterXML/jackson-databind/pull/5175/commits/eaf80fd86805f6b624368a817b5d18e81b1cc16f). My first idea was to keep track of already injected properties leveraging a flag in `SettableBeanProperty`, but I think [you'd like to avoid that](https://github.com/FasterXML/jackson-databind/pull/5224#issuecomment-3076722020):
> [...] although would prefer not adding configuring state in AnnotatedMember

Then, I tried leveraging the existing parameters such as `_paramsNeeded` and `_paramsSeen` in `PropertyValueBuffer`, but they're not good for this case, at least not as they're currently used.

So, I added `_paramsInjected` and `_paramsInjectedBig` to keep track of injected parameters in a similar but dedicated way.

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Minor javadoc fix

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Fix name reference

cowtowncoder committed:
Improve test to verify injection field name

cowtowncoder committed:
Remove unnecessary diff

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/PropertyValueBuffer.java:
```diff
@@ -302,6 +336,32 @@ protected Object _findMissing(SettableBeanProperty prop) throws JsonMappingExcep
         }
     }
 
+    private void _inject(final SettableBeanProperty prop) throws JsonMappingException {
```
Needs comment/javadoc: I think this is only called for cases where we post-process "missing" Creator properties: ones for which no Input value was provided.

cowtowncoder replied:
... Actually, no, scratch that. Figured it out.

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/PropertyValueBuffer.java:
```diff
@@ -228,6 +248,19 @@ public Object[] getParameters(SettableBeanProperty[] props)
             }
         }
+
+        if (_paramsInjectedBig == null) {
```
Ok so now my remaining implementation problem is this: we should not

1. Allocate `_paramsInjectedBig` if no injectable values are defined (to avoid overhead for common enough case) OR
2. Do post-processing similarly (scanning over all properties for "small" case).

because I think it's not good to impose measurable overhead for non-injection cases.

To avoid these, checking for "are there any Injectables" would need to be done by pre-processing `SettableBeanProperty` set, I think, and passing a flag or something in `PropertyValueBuffer` constructor, because it cannot be done efficiently in `getParameters()`.
Based on this flag, could eagerly construct `_paramsInjectedBig`.

I realize this is not trivial to do... but I feel it is important for performance.

As an extension, could create new helper class for holding Injectable status (`BufferInjectables` inner class?), and its existence could be used instead of flag.

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/deser/impl/PropertyValueBuffer.java:
```diff
@@ -228,6 +248,19 @@ public Object[] getParameters(SettableBeanProperty[] props)
+
+        if (_paramsInjectedBig == null) {
+            for (int ix = 0; ix < _creatorParameters.length; ++ix) {
+                if ((_paramsInjected & 1) == 0) {
```
Bug! (thanks Claude :) ) -- needs to be

    if ((_paramsInjected & (1 << ix)) == 0) {

Will fix.

cowtowncoder committed:
Fix a bug

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Add parentheses for clarity

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder committed:
Optimize tracking of injectable values

cowtowncoder committed:
...

cowtowncoder committed:
...

cowtowncoder committed:
...

cowtowncoder committed:
Merge branch '2.x' into feature/databind-1381-inject-only

cowtowncoder reviewed (approved):
LGTM

cowtowncoder commented:
@giulong Phew! Took a while, but I was able to make changes that should reduce overhead of injection by pre-calculating indexes of properties and using that to avoid unnecessary scanning.

I am ready to FINALLY merge this, but thought you might want to have a look first to sanity check my changes?

cowtowncoder commented on src/main/java/com/fasterxml/jackson/databind/DeserializationContext.java:
```diff
@@ -469,10 +469,12 @@ public final Object findInjectableValue(Object valueId,
+                    || Boolean.TRUE.equals(optional)
+                    || ((useInput == null || optional == null)
+                            && !isEnabled(DeserializationFeature.FAIL_ON_UNKNOWN_INJECT_VALUE))) {
                 return JacksonInject.Value.empty();
```
Hmmmh. This is wrong -- should be `null`, its Injectable value, NOT injection definition.

cowtowncoder replied:
Changed in 2.20, looks like; quietly returning new marker value. Should have caught this problem back then; I think I'll this now for 2.21.
